Move hal stubs to hal directory

rajkan01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
fixes #14789 

- This PR makes the mbed-stubs-hal library more self-contained by removing the dependencies on the monolithic header libraries we created to make implementing the stubs easier. We also move the files to the hal directory so the stubs live with the associated component.
- move mbed-headers-hal library creation inside doubles CMake.


<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
None.
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
None
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
To be updated
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

rajkan01 committed:
CMake: Move hal stubs sources to the mbed-os/hal directory

Move the hal stubs into the hal component directory.
so we can avoid duplicating the mbed-os source tree in a central
UNITTESTS folder.

rajkan01 committed:
CMake: unit-tests: Move mbed-headers-hal to the hal directory

Move the header-only mbed-headers-hal library the unit test stubs
depend on into the hal component directory. This makes the events
stubs more self-contained and improves the composition of the library.

rajkan01 committed:
CMake: unit-tests: Make hal stubs only depend on headers it uses

Previously the hal stub library depended on `mbed-headers`, which
is a collection of all available headers in mbed-os. To make it easier
to separate the library, only depend on the headers we're using.

LDong-Arm reviewed (approved):
LGTM

LDong-Arm commented on hal/tests/UNITTESTS/doubles/CMakeLists.txt:
```diff
@@ -24,6 +32,8 @@ target_link_options(mbed-stubs-hal
 
```
Given that `--coverage` is globally set in UNITTESTS/CMakeLists.txt, can we remove it from here (the four lines above this)? I also asked in #14773 about the platform stubs - we can investigate and align them.

rajkan01 replied:
Initially, we added target_link_libraries mbed-stubs-hal with gcov to resolve the dependency issue related to coverage, Rob was testing unit test changes with Armclang (build failure), it didn't work, so he proposed to use the target link option with --coverage to resolve that issue refer commit  https://github.com/ARMmbed/mbed-os/pull/14426/commits/10625b35b1818520deb5cd51369265fcffcc4022 , maybe we need to remove these changes and try with Armclang

rajkan01 replied:
I have removed --coverage link options changes and tested with ubuntu Linux , please re-review

rajkan01 committed:
CMake: Remove --coverage link option from mbed-stubs-hal

The --coverage flag globally set in UNITTESTS/CMakeLists.txt
and configuring the same option again in mbed-stubs-hal is
redundant so remove it

LDong-Arm reviewed (approved):
LGTM, just need to rerun CI
